aws/resource_aws_gamelift_game_session_queue.go

```diff
@@ -0,0 +1,204 @@
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 128),
+			},
+			"player_latency_policy": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"maximum_individual_player_latency_milliseconds": {
+							Type:         schema.TypeInt,
+							Required:     true,
+							ValidateFunc: validation.IntAtLeast(0),
+						},
+						"policy_duration_seconds": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							ValidateFunc: validation.IntAtLeast(0),
+						},
+					},
+				},
+			},
+			"timeout_in_seconds": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				ValidateFunc: validation.IntBetween(10, 600),
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsGameliftGameSessionQueueCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).gameliftconn
+
+	input := gamelift.CreateGameSessionQueueInput{
+		Name:                  aws.String(d.Get("name").(string)),
+		Destinations:          expandGameliftGameSessionQueueDestinations(d.Get("destinations").([]interface{})),
+		PlayerLatencyPolicies: expandGameliftGameSessionPlayerLatencyPolicies(d.Get("player_latency_policy").([]interface{})),
+		TimeoutInSeconds:      aws.Int64(int64(d.Get("timeout_in_seconds").(int))),
+	}
+	log.Printf("[INFO] Creating Gamelift Session Queue: %s", input)
+	out, err := conn.CreateGameSessionQueue(&input)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*out.GameSessionQueue.GameSessionQueueArn)
+	d.Set("name", out.GameSessionQueue.Name)
+
+	return resourceAwsGameliftGameSessionQueueRead(d, meta)
+}
+
+func resourceAwsGameliftGameSessionQueueRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).gameliftconn
+	log.Printf("[INFO] Describing Gamelift Session Queues: %s", d.Get("name"))
+	limit := int64(1)
+	out, err := conn.DescribeGameSessionQueues(&gamelift.DescribeGameSessionQueuesInput{
+		Names: aws.StringSlice([]string{d.Get("name").(string)}),
+		Limit: &limit,
+	})
+	if err != nil {
+		if isAWSErr(err, gamelift.ErrCodeNotFoundException, "") {
+			log.Printf("[WARN] Gamelift Session Queues (%s) not found, removing from state", d.Get("name"))
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+	sessionQueues := out.GameSessionQueues
+
+	if len(sessionQueues) < 1 {
+		log.Printf("[WARN] Gamelift Session Queue (%s) not found, removing from state", d.Get("name"))
+		d.SetId("")
+		return nil
+	}
+	if len(sessionQueues) != 1 {
+		return fmt.Errorf("expected exactly 1 Gamelift Session Queues, found %d under %q",
+			len(sessionQueues), d.Get("name"))
+	}
+	sessionQueue := sessionQueues[0]
+
+	d.Set("arn", sessionQueue.GameSessionQueueArn)
+	d.Set("name", sessionQueue.Name)
+	d.Set("timeout_in_seconds", sessionQueue.TimeoutInSeconds)
+	d.Set("destinations", sessionQueue.Destinations)
```

Comment:
When using `d.Set()` with aggregate types (TypeList, TypeSet, TypeMap), [we should perform error checking to prevent issues where the code is not properly able to set the Terraform state](https://www.terraform.io/docs/extend/best-practices/detecting-drift.html#error-checking-aggregate-types). e.g.

```go
if err := d.Set("destinations", sessionQueue.Destinations); err != nil {
  return fmt.Errorf("error setting destinations: %s", err)
}
```

Without this check and if there is a type conversion error, Terraform is silently accepting the configuration "as-is" into the Terraform state and unable to detect drift from the API should changes occur outside Terraform.

After adding this error checking, you should start receiving type errors as `sessionQueue.Destinations` is a `[]*GameSessionQueueDestination` SDK object and not a `[]interface{}`. We generally handle this via a "flattenServiceObject()" function. e.g.

```go
func flattenGameliftGameSessionQueueDestinations(destinations []*gamelift.GameSessionQueueDestination) []interface{} {
	l := make([]interface{}, 0)

	for _, destination := range destinations {
		if destination == nil {
			continue
		}
		l = append(l, aws.StringValue(destination.DestinationArn))
	}

	return l
}
```

Which can be referenced via:

```go
if err := d.Set("destinations", flattenGameliftGameSessionQueueDestinations(sessionQueue.Destinations)); err != nil {
  return fmt.Errorf("error setting destinations: %s", err)
}
```

Reply:
Just curious - while we should skip `nil` values in `destinations`? How they can get there? Why it is not the error if we got such destination?

Reply:
This is just a safety check in case the API does something unexpected. Terraform will completely stop running (panic) if you try to perform a dereference against a `nil` value.